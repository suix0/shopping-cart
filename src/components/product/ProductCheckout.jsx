import InputField from "../../layouts/modal/InputField";
import PropTypes from "prop-types";
// import del from ".../assets/delete.svg"

const ProductCheckout = ({
  cartProducts,
  setProductQuantity,
  deleteProduct,
}) => {
  // const updateProductQuantity = (quantity) => {
  //   setProductQuantity(quantity);
  //   localStorage.setItem("cartProducts", JSON.stringify(cartProducts));
  // };

  return (
    <div className="flex justify-between">
      <div className="flex gap-4">
        <img
          src={cartProducts.product.image}
          alt={cartProducts.product.title}
          className="w-24"
        />
        <div>
          <p>{cartProducts.product.title}</p>
          <p>${cartProducts.product.price}</p>
          <p className="font-thin text-md">Quantity</p>
          <div className="flex items-center">
            <InputField
              productQuantity={cartProducts.quantity}
              setProductQuantity={setProductQuantity}
              isCheckout={true}
              id={cartProducts.product.id}
              itemId={cartProducts.id}
              deleteProduct={deleteProduct}
            ></InputField>
          </div>
        </div>
      </div>
      <p>
        ${`${(cartProducts.product.price * cartProducts.quantity).toFixed(2)}`}
      </p>
    </div>
  );
};

ProductCheckout.propTypes = {
  cartProducts: PropTypes.object,
  product: PropTypes.shape({
    title: PropTypes.string,
    price: PropTypes.number,
    quantity: PropTypes.number,
  }),
  setProductQuantity: PropTypes.func,
  deleteProduct: PropTypes.func,
};

export default ProductCheckout;
