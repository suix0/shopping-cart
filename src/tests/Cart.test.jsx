import { render, screen, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi, describe, it, expect, afterAll } from "vitest";
import ProductCheckout from "../components/product/ProductCheckout";

// setup props
const cartProducts = {
  product: {
    id: 1,
    title: "Mock 1",
    price: 15.99,
  },
  quantity: 5,
  id: crypto.randomUUID(),
};

const setProductQuantity = vi.fn((e) => {
  if (e.target.textContent === "+") {
    cartProducts.quantity += 1;
  } else if (e.target.textContent === "-") {
    cartProducts.quantity -= 1;
  }
});

describe("product checkout", () => {
  afterAll(() => {
    cleanup();
  });

  it("renders product checkout", () => {
    const { container } = render(
      <ProductCheckout cartProducts={cartProducts} />
    );

    expect(container).toMatchSnapshot();
  });

  // Test whether total price of a product updates relative to quantity
  it("increases quantity should change total price of product accordingly", async () => {
    const user = userEvent.setup();

    const { rerender } = render(
      <ProductCheckout
        cartProducts={cartProducts}
        setProductQuantity={setProductQuantity}
      />
    );

    const increaseButton = screen.getByRole("button", { name: "+" });

    await user.click(increaseButton);

    rerender(
      <ProductCheckout
        cartProducts={cartProducts}
        setProductQuantity={setProductQuantity}
      />
    );

    const input = screen.getByRole("textbox");

    expect(input.value).toBe("6");
    expect(screen.getByText("$95.94")).toBeInTheDocument(); // price * updated quantity
  });

  // Test whether total price of a product updates relative to quantity
  it("decreases quantity should change total price of product accordingly", async () => {
    const user = userEvent.setup();

    const { rerender } = render(
      <ProductCheckout
        cartProducts={cartProducts}
        setProductQuantity={setProductQuantity}
      />
    );

    const decreaseButton = screen.getByRole("button", { name: "-" });

    await user.click(decreaseButton);

    rerender(
      <ProductCheckout
        cartProducts={cartProducts}
        setProductQuantity={setProductQuantity}
      />
    );

    const input = screen.getByRole("textbox");

    expect(input.value).toBe("5");
    expect(screen.getByText("$79.95")).toBeInTheDocument(); // price * updated quantity
  });
});
