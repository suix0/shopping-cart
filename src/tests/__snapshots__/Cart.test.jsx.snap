// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`product checkout > renders product checkout 1`] = `
<div>
  <div>
    <img
      alt=""
      src=""
    />
    <div>
      <div>
        <p>
          Mock 1
        </p>
        <p>
          $
          15.99
        </p>
        <div>
          <div
            id="1"
          >
            <p>
              Quantity
            </p>
            <div
              class="inline-flex border justify-around w-20 p-[4px]"
            >
              <button
                class="false select-none"
                id="1"
              >
                -
              </button>
              <input
                class="w-full text-center outline-none"
                id="1"
                type="text"
                value="5"
              />
              <button
                class="select-none"
                id="1"
              >
                +
              </button>
            </div>
          </div>
          <img
            alt=""
            src=""
          />
        </div>
      </div>
      <p>
        $
        79.95
      </p>
    </div>
  </div>
</div>
`;
